staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/compilation.go

```diff
@@ -31,11 +32,25 @@ const ScopedTypeName = "apiextensions.k8s.io.v1alpha1.ValidationExpressionSelf"
 // ScopedVarName is the variable name assigned to the locally scoped data element of a CEL valid.
 const ScopedVarName = "self"
 
+// CompilationResults represents the compilation results got from cel compilation
+type CompilationResults struct {
+	Results []CompilationResult
+	Error   error
+}
+
+// CompilationResult represents the cel compilation result for one rule
+type CompilationResult struct {
+	Rule      apiextensions.ValidationRule
+	Program   cel.Program
+	Errors    []error
+	RuleIndex int
+}
+
 // Compile compiles all the CEL validation rules in the CelRules and returns a slice containing a compiled program for each provided CelRule, or an array of errors.
-func Compile(s *schema.Structural) ([]cel.Program, []error) {
-	var allErrors []error
+func Compile(s *schema.Structural) CompilationResults {
```

Comment:
what about `([]compilationResult, error)`? Just to avoid structs with similar names.